### ISSUE TITLE
Consider refactoring impact on service clients

### ISSUE THREAD

**Issue by SpencerArtisan** [leonjones1974/shimrr#2]
In the tutorial you have 

```
trait ProductService {
  def allProducts(): Iterable[BicycleV4]
}
```

Which creates a refactoring impact on every client to this service.  Would be good if the concept of versions was totally hidden behind the service, so it remained as

```
trait ProductService {
  def allProducts(): Iterable[Bicycle]
}
```

Not sure how you'd achieve this.  One option would be to have the usual BicycleV1, V2, V3, etc, with Bicycle essentially being an alias to the latest version.  This would bugger up your use of Bicycle as a trait though.  You could just rename the trait to BicycleProduct or something.

**Comment by leonjones1974** [comment]
Yes, in reality they are two different things.  One being the representation that has been read from the data source and the other the model destined for consumers.

Agree is would be nice for some changes to be non breaking

Still not convinced it requires any code in the library, but at the minimum I'll add it to the tutorial as a recommended pattern.  Type aliasing might be a good solution here

**Comment by SpencerArtisan** [comment]
I definitely see this as a recommended pattern rather than a code change
